pipeline/relation_embedder/src/main/scala/uk/ac/wellcome/relation_embedder/RelationEmbedderWorkerService.scala

```diff
@@ -48,16 +48,19 @@ class RelationEmbedderWorkerService[MsgDestination](
         relationsService.getRelations(work).map(work.transition[Denormalised])
       }
       .groupedWithin(batchSize, flushInterval)
-      .mapAsync(2)(workIndexer.index)
-      .collect { case Left(failedWorks) => failedWorks.toList }
+      .mapAsync(2) { work =>
+        workIndexer.index(work).flatMap {
+          case Left(failedWorks) =>
+            Future.failed(
+              new Exception(s"Failed indexing works: $failedWorks")
+            )
+          case Right(works) => Future.successful(works.toList)
+        }
+      }
       .mapConcat(identity)
-      .toMat(Sink.seq)(Keep.right)
-      .run()
-      .flatMap {
-        case Nil => Future.successful(())
-        case failedWorks =>
-          Future.failed(
-            new Exception(s"Failed indexing works: $failedWorks")
-          )
+      .mapAsync(2) { work =>
```

Comment:
Does this need `mapAsync() { _ => Future.fromTry(...) }`? `Future.fromTry` returns an already completed future (ie nothing here is async) so I think you should be able to throw the `Failure(...)` directly without redundantly wrapping it

Reply:
Good point :+1: